### PR TITLE
Screen mirror

### PR DESCRIPTION
**Pull request by StijnOostdam**
Added screen mirror functionality

### CODE REVIEW

File: Nanoleaf-Models/Enums/ScreenMirrorAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Winleafs.Models.Enums
+{
+    public enum ScreenMirrorAlgorithm
+    {
+        Ambilight = 0,
+        ScreenMirror = 1,
+        //NearestNeighbour = 2
```
Comment (bartdebever) [inline Nanoleaf-Models/Enums/ScreenMirrorAlgorithm.cs]: commented out code

File: Nanoleaf-wpf/Api/Effects/ScreenMirrorEffects/ScreenMirror.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            _nanoleafClient = nanoleafClient;
+
+            _externalControlEndpoint = _nanoleafClient.ExternalControlEndpoint;
+
+            _screenBounds = MonitorHelper.GetScreenBounds(device.ScreenMirrorMonitorIndex);
+
+            _panels = orchestrator.PanelLayout.GetScaledTriangles(_screenBounds.Width, _screenBounds.Height);
+
+            //Set the rectangle size to 1/3th of the length of a side of the triangle. TODO: test what size is best
```
Comment (bartdebever) [inline Nanoleaf-wpf/Api/Effects/ScreenMirrorEffects/ScreenMirror.cs]: TODO?

File: Nanoleaf-wpf/Api/Layouts/PanelLayout.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+using Winleafs.Models.Models.Layouts;
+
+namespace Winleafs.Wpf.Api.Layouts
+{
+    public class PanelLayout
+    {
+        private static readonly SolidColorBrush _borderColor = (SolidColorBrush)Application.Current.FindResource("NanoleafBlack");
+        private static int _defaultTriangleSize = 148; //The default size in pixels of a triangle side as reported by the Nanoleaf API
+
+        private INanoleafClient _nanoleafClient;
```
Comment (bartdebever) [inline Nanoleaf-wpf/Api/Layouts/PanelLayout.cs]: Make this readonly

File: Nanoleaf-wpf/Api/Layouts/PanelLayout.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+            foreach (var triangle in _unscaledTriangles)
+            {
+                //Move MidPoint
+                triangle.MidPoint = new Point(triangle.MidPoint.X - minTriangleX, triangle.MidPoint.Y - minTriangleY);
+
+                //Move triangle
+                for (var i = 0; i < triangle.Polygon.Points.Count; i++)
+                {
+                    var x = triangle.Polygon.Points[i].X - minTriangleX;
```
Comment (bartdebever) [inline Nanoleaf-wpf/Api/Layouts/PanelLayout.cs]: Is `-=` an option with it

File: Nanoleaf-wpf/Helpers/ScreenGrabber.cs [hunk 5]
```diff
@@ -1,137 +1,64 @@
...
-
-            var monitorInfo = new MonitorInfo();
-            EnumDisplaySettings(Screen.AllScreens[UserSettings.Settings.AmbilightMonitorIndex].DeviceName, -1, ref monitorInfo);
-
-            _screenBounds = new Rectangle(monitorInfo.dmPositionX, monitorInfo.dmPositionY, monitorInfo.dmPelsWidth, monitorInfo.dmPelsHeight);
-        }
-#pragma warning restore S3963 // "static" fields should be initialized inline
-
-        private static Bitmap CaptureScreen()
+        public static Bitmap CaptureScreen(Rectangle screenBounds)
```
Comment (bartdebever) [inline Nanoleaf-wpf/Helpers/ScreenGrabber.cs]: General idea. What if I set my monitor 2 as my screen mirror and remove the screen from my computer afterwards. Will that cause a crash?

File: Nanoleaf-wpf/ViewModels/OptionsViewModel.cs [hunk 6]
```diff
@@ -1,25 +1,129 @@
...
+            get => _screenMirrorControlBrightness;
+            set
+            {
+                _screenMirrorControlBrightness = value;
+                ScreenMirrorControlBrightnessPerDevice[_selectedDevice] = _screenMirrorControlBrightness;
+            }
+        }
+
+        #region Screen mirror algorithm dropdown
+        private Dictionary<string, ScreenMirrorAlgorithm> _screenMirrorAlgorithmMapping { get; set; } //Map display values to enum values
```
Comment (bartdebever) [inline Nanoleaf-wpf/ViewModels/OptionsViewModel.cs]: General readability would improve if this comment is put on the line above (falls off the screen on github)

File: Nanoleaf-wpf/ViewModels/OptionsViewModel.cs [hunk 7]
```diff
@@ -1,25 +1,129 @@
...
+        public string SelectedScreenMirrorAlgorithm
+        {
+            get { return EnumLocalizer.GetLocalizedEnum(_selectedScreenMirrorAlgorithm); }
+            set
+            {
+                _selectedScreenMirrorAlgorithm = _screenMirrorAlgorithmMapping[value];
+
+                AlgorithmPerDevice[_selectedDevice] = _selectedScreenMirrorAlgorithm;
+
+                //_parent.ScreenMirrorAlgorithmChanged(); Will be used later if we visualize the algorithms
```
Comment (bartdebever) [inline Nanoleaf-wpf/ViewModels/OptionsViewModel.cs]: Is this still a task to be done?

File: Nanoleaf-wpf/ViewModels/OptionsViewModel.cs [hunk 8]
```diff
@@ -1,25 +1,129 @@
...
+            }
+        }
+
+        public IEnumerable<string> ScreenMirrorAlgorithms
```
Comment (bartdebever) [inline Nanoleaf-wpf/ViewModels/OptionsViewModel.cs]: public IEnumerable<string> ScreenMirrorAlgorithms => _screenMirrorAlgorithmMapping.Keys;

File: Nanoleaf-wpf/ViewModels/OptionsViewModel.cs (hunk 7 above)
Comment (StijnOostdam) [reply Nanoleaf-wpf/ViewModels/OptionsViewModel.cs]: Yes, part of a future feature

File: Nanoleaf-wpf/Api/Effects/ScreenMirrorEffects/ScreenMirror.cs (hunk 2 above)
Comment (StijnOostdam) [reply Nanoleaf-wpf/Api/Effects/ScreenMirrorEffects/ScreenMirror.cs]: Intentional, is part of a task in the project.

File: Nanoleaf-wpf/Api/Layouts/PanelLayout.cs (hunk 4 above)
Comment (StijnOostdam) [reply Nanoleaf-wpf/Api/Layouts/PanelLayout.cs]: No, they are not modifiable at that point

File: Nanoleaf-wpf/Helpers/ScreenGrabber.cs (hunk 5 above)
Comment (StijnOostdam) [reply Nanoleaf-wpf/Helpers/ScreenGrabber.cs]: Good one, added checks in both the options view and the effects to catch this rare behaviour.

**Pull request merged**